include/aie/Dialect/AIEVec/Transforms/FoldMulAddChainToConvOp.h

```diff
@@ -0,0 +1,511 @@
+      // Transfer
+      // %c32 = arith.constant 32 : index
+      // %1 = aievec.upd %arg1[%c32] {index = 0 : i8} : vector<32xi8>
+      // %2 = aievec.broadcast %1 {idx = 0 : i8} : vector<32xi8>
+      // to -
+      // %c0 = arith.constant 0 : index
+      // %1 = aievec.upd %arg1[%c0] {index = 0 : i8} : vector<64xi8>
+      // %2 = aievec.broadcast %1 {idx = 32 : i8} : vector<32xi8>
+      if (auto idxDefOp = innerMostIdx.getDefiningOp()) {
+        if (auto constOp = dyn_cast<arith::ConstantOp>(idxDefOp)) {
+          val = constOp.getValue().cast<IntegerAttr>().getInt();
+          if (val) {
+            defIdx = val / lanes * lanes;
+            val %= lanes;
+            newIdx = rewriter.create<arith::ConstantOp>(
+                constOp.getLoc(),
+                rewriter.getIntegerAttr(constOp.getType(), defIdx));
+            indices[indices.size() - 1] = newIdx;
+          }
+        }
+      }
+
+      aievec::UPDOp newBcastOp = bcastUPDOp;
+
+      // Rewrite the upd op with maximum vector lanes
+      if (vType != resType) {
+        newBcastOp = rewriter.create<aievec::UPDOp>(
+            bcastUPDOp->getLoc(), resType, bcastUPDOp.getSource(), indices, 0,
+            0, TypedValue<VectorType>(nullptr));
+      }
+
+      // Since we do not need to use duplicated data like in AIE1, if a
+      // dup-factor exists, we extract the identical data by shuffle op. We use
+      // mode 0 to extract the elements with even indices for i8 type data.
+      Operation *shuffleOp = newBcastOp;
+      if (dupFactor != 1) {
+        shuffleOp = rewriter.create<aievec::ShuffleOp>(
+            newBcastOp.getLoc(), resType, newBcastOp.getResult(), 0);
+      }
+
+      int32_t shiftBytes = (bcastOp.getIdx() + val) *
+                           getElementSizeInBits(vType) / 8 / dupFactor;
+
+      rhs = shuffleOp->getResult(0);
+
+      // Generate a shift_bytes operation for rhs if the start position is not
+      // 0.
+      if (shiftBytes) {
+        SmallVector<Value> sources = {shuffleOp->getResult(0)};
+
+        rhs = rewriter.create<aievec::ShiftOp>(
+            shuffleOp->getLoc(), sources.back().getType().cast<VectorType>(),
+            sources, shiftBytes);
+      }
+
+      aievec::UPDOp lUPDOp = cast<aievec::UPDOp>(lhs.getDefiningOp());
+      SmallVector<Value, 8> lIndices;
+      lIndices.append(lUPDOp.getIndices().begin(), lUPDOp.getIndices().end());
+
+      lhs = rewriter.create<aievec::UPDOp>(lUPDOp->getLoc(), resType,
+                                           lUPDOp.getSource(), lIndices, 0, 0,
+                                           TypedValue<VectorType>(nullptr));
+
+      auto notMulOrAddOp = [&](Operation *op) {
+        return !isa<arith::MulIOp, arith::AddIOp>(op);
+      };
+
+      Operation *convOp = nullptr;
+      arith::AddIOp lastAddOp =
+          cast<arith::AddIOp>(*(fusedOps.back()->getUsers().begin()));
+
+      if (llvm::any_of(lastAddOp->getUsers(), notMulOrAddOp)) {
```

Comment:
I would expected that we care only whether the last add has a user that is not an add.
Why do you instead check whether the last add has a user that is not a add and not a mul?